[WPE] WPE Platform: add API to handle visibility

carlosgcampos opened the pull request:
#### 37eef8a67009b9162a9be622ad4a8c429331982b
<pre>
[WPE] WPE Platform: add API to handle visibility
<a href="https://bugs.webkit.org/show_bug.cgi?id=275482">https://bugs.webkit.org/show_bug.cgi?id=275482</a>

Reviewed by Nikolas Zimmermann.

Users should be able to make a view visible/hidden and platform
implementations should map/unmap the view according to the toplevel
state.

* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::m_backend):
* Source/WebKit/UIProcess/API/wpe/qt6/WPEQtView.cpp:
(WPEQtView::createWebView):
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpeViewGetProperty):
(wpe_view_class_init):
(wpe_view_get_visible):
(wpe_view_set_visible):
(wpe_view_get_mapped):
(wpe_view_map):
(wpe_view_unmap):
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
* Source/WebKit/WPEPlatform/wpe/drm/WPEViewDRM.cpp:
(wpeViewDRMConstructed):
* Source/WebKit/WPEPlatform/wpe/headless/WPEViewHeadless.cpp:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(wpeViewWaylandConstructed):
(wpeViewWaylandCanBeMapped):
(wpe_view_wayland_class_init):

Canonical link: <a href="https://commits.webkit.org/280113@main">https://commits.webkit.org/280113@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3c33ecbb654c5a0af94d3a0394a54ab48bf3dbf7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55683 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35006 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8150 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58667 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6113 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42628 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6312 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44844 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4208 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57712 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32905 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47994 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25976 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29692 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4257 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51678 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5590 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60258 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5720 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52273 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48064 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51762 "Passed tests") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8230 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30837 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31922 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33003 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31669 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/WPEView.h:
```diff
@@ -119,6 +120,12 @@ WPE_API void                    wpe_view_resized                       (WPEView
+WPE_API gboolean                wpe_view_get_visible                   (WPEView      *view);
+WPE_API void                    wpe_view_set_visible                   (WPEView      *view,
+                                                                        gboolean      visible);
+WPE_API gboolean                wpe_view_get_mapped                    (WPEView      *view);
```
I think the names are too generic, and thus confusing. Reading only `wpe_view_get_mapped` - I would have no idea what it does...

nikolaszimmermann replied:
Since I cannot propose anything better _and_ it's existing X11/Gtk terminology, let's stick with it.

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/WPEView.h:
```diff
@@ -119,6 +120,12 @@ WPE_API void                    wpe_view_resized                       (WPEView
+WPE_API void                    wpe_view_set_visible                   (WPEView      *view,
+                                                                        gboolean      visible);
+WPE_API gboolean                wpe_view_get_mapped                    (WPEView      *view);
+WPE_API void                    wpe_view_map                           (WPEView      *view);
```
Same for wpe_view_map/unmap.

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -556,6 +596,106 @@ void wpe_view_scale_changed(WPEView* view, gdouble scale)
+ * wpe_view_unmap:
+ * @view: a #WPEView
+ *
+ * Make @view to be unmapped.
```
What?! :-) This doesn't resolve my confusion.

nikolaszimmermann commented on Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
```diff
@@ -94,6 +93,9 @@ View::View(struct wpe_view_backend* backend, WPEDisplay* display, const API::Pag
 #endif
 
+    if (m_backend)
+        m_viewStateFlags = { WebCore::ActivityState::WindowIsActive, WebCore::ActivityState::IsFocused, WebCore::ActivityState::IsVisible, WebCore::ActivityState::IsInWindow };
```
Why is this only set if m_backend is non-null?

carlosgcampos replied:
Because that's old api.

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -51,6 +51,8 @@ struct _WPEViewPrivate {
     WPEViewState state;
     bool closed;
+    bool visible { true };
+    bool mapped;
```
Shouldn't we default to false here? (ah c struct :-)

carlosgcampos replied:
The view is visible by default

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -237,6 +250,33 @@ static void wpe_view_class_init(WPEViewClass* viewClass)
+    /**
+     * WPEView:mapped:
+     *
+     * Whether the view is mapped or not. A view is mapped when #WPEView:visible
```
I'm confused. For PROP_VISIBLE you mention, a view might not be mapped even if it's set as visible (e.g. for minimized windows). PROP_MAPPED says a view is mapped if it's visible, and not hidden (e.g. minimized). That's a contradiction.

carlosgcampos replied:
I don't see the contradiction, if a view toplevel is minimized, in a different workspace, the screen is locked, etc. the view is never mapped, no matter what the visibility property says. But to be mapped it also needs to be visible.

nikolaszimmermann replied:
I think the wording was just difficult to interpret - much clearer now.

nikolaszimmermann commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -384,8 +384,10 @@ static const struct wl_surface_listener surfaceListener = {
         wpeViewWaylandUpdateScale(view);
-        if (monitorChanged)
+        if (monitorChanged) {
+            wpe_view_map(WPE_VIEW(view));
```
Don't we need to unmap the "previously" mapped view?

carlosgcampos replied:
We don't support multiple view yet, so there's no previously mapped view.

nikolaszimmermann reviewed (approved):
Great, thanks for updating the documentation.

carlosgcampos committed:
[WPE] WPE Platform: add API to handle visibility
https://bugs.webkit.org/show_bug.cgi?id=275482

Reviewed by Nikolas Zimmermann.

Users should be able to make a view visible/hidden and platform
implementations should map/unmap the view according to the toplevel
state.

* Source/WebKit/UIProcess/API/wpe/WPEWebView.cpp:
(WKWPE::m_backend):
* Source/WebKit/UIProcess/API/wpe/qt6/WPEQtView.cpp:
(WPEQtView::createWebView):
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpeViewGetProperty):
(wpe_view_class_init):
(wpe_view_get_visible):
(wpe_view_set_visible):
(wpe_view_get_mapped):
(wpe_view_map):
(wpe_view_unmap):
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
* Source/WebKit/WPEPlatform/wpe/drm/WPEViewDRM.cpp:
(wpeViewDRMConstructed):
* Source/WebKit/WPEPlatform/wpe/headless/WPEViewHeadless.cpp:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(wpeViewWaylandConstructed):
(wpeViewWaylandCanBeMapped):
(wpe_view_wayland_class_init):

Canonical link: https://commits.webkit.org/280113@main
